Add includeDOM option to rule "forbid-component-props".

EdwardDrapkin opened the pull request:
I wanted to forbid some prop types on DOM nodes as well as Components, so I added the option here.  Let me know if there's anything that needs to be done here to get this approved.  Thanks in advance!

Commit:
first try to includeDOM option

ljharb commented on docs/rules/forbid-component-props.md:
```diff
@@ -35,10 +35,14 @@ The following patterns are not considered warnings:
 
 An array of strings, with the names of props that are forbidden. The default value of this option is `['className', 'style']`.
+
+### `includeDOM`
```
This seems like something that will vary per prop name. Perhaps the enhancement should be that `forbid` can have strings or objects, and if an object, it can be something like `{ "property": "foo", "includeDOM": true }`?

ljharb reviewed (changes requested):
Either way, this will need tests.

EdwardDrapkin commented:
How about `forbid` and `forbidOnDOM`?  `forbidOnDOM` would be an array of strings just like `forbid` except that it's empty by default.

ljharb commented:
I think that allowing string-or-object in the array gives far greater flexibility - that would also let me potentially forbid a prop on one list of components, but not on others.

EdwardDrapkin commented:
I think the flexibility is the same: 

     {
      "forbid":  ["className", "style"], 
      "forbidOnDOM": ["style"] 
     }  

would achieve the same thing, no?

I think that's easier for users to manage than:  `

       {
         "forbid": [ "style", { "className": { "forbidOnDOM": true } } ]
       }

Or
       {
         "forbid": [ "style", { "property": "className", "forbidOnDOM": true } ]
       }

It's a completely subjective decision and ultimately it's your code, so I'll do what you think is best, but IMO the first example is the easiest to read if I'm a hypothetical engineer and I'm reading someone else's .eslintrc and I'm not necessarily familiar with this particular plugin.

ljharb commented:
@EdwardDrapkin imagine a hypothetical future option where you could forbid "foo" on a Foo component, but allow "foo" on anything else - or where you want to forbid "style" on both React and DOM components, but simultaneously forbid "className" only on React components.

The latter certainly can be achieved with your suggestion, but that won't allow for the former, and that also requires that property names be duplicated in both lists, which can be hard to maintain.

EdwardDrapkin commented:
Ahhhhh, I wasn't even considering that!  You make excellent points.

Perhaps we could also take this opportunity to allow people the flexibility you imagine?  I have two ideas.

First, we could allow passing custom validator functions like:

      {
        "property": "style",
        "validator": "no-dom-components"|"only-dom-components"|"all-components"|function 
      }

where the function is just `(componentName:string) => boolean` that the plugin would use to choose whether or not to apply the filters to the component?  Alternatively, I could do `(componentName:string, property:string) => boolean` and expose the validation step itself to the user, but I'm having trouble imagining a case where that more useful than the former. 

This would be pretty simple to implement, and is probably the most flexible approach, but not exactly user friendly because it's too complex for what I believe is likely to be the common use case...

So maybe, instead we just allow a regular expression to be passed?  

      {
        "property": "style",
        "isValid": "^[A-Z]" 
      }

This is probably the easiest for users to use and covers like 99% of the cases people would want, but would require documentation to tell people to use something like `isValid: "."` to match all components' names, or a special case.

What do you think?

ljharb commented:
eslint configs are defined in JSON or yaml, so it'd be impossible to provide a function.

I think adding regex support is unneeded complexity - being able to configure a boolean for DOM and another for React, along with simple component names per-property that can override the booleans, seems like it would cover all the current and potential use cases pretty well.

EdwardDrapkin commented:
Can't you use a JS file as an .eslintrc.js?  It's irrelevant anyway...

How does this look?

     {
       "property": <string>,
       "include": "React"|"DOM"|"Both"|<array> 
     }

Is `include` a good param name?

ljharb commented:
I think I'd prefer:
```js
{
  "property": <string>,
  "allowOnComponents": <boolean>, // default false
  "allowOnDOM": <boolean>, // default false
  "forbid": <array of component name strings>, // overrides previous two booleans
}
```
where passing a simple string "foo" as the property name is equivalent to:
```js
{
    "property": "foo",
    "allowOnComponents": false,
    "allowOnDOM": true, // simply so it's not a breaking change
    "forbid": [],
}
```

EdwardDrapkin commented:
Perfect.  I'll get this PR updated sometime in the next 24-48 hours with these changes!

Commit:
New API per request + tested

ljharb commented on docs/rules/forbid-component-props.md:
````diff
@@ -35,10 +37,94 @@ The following patterns are not considered warnings:
 ```js
 ...
-"forbid-component-props": [<enabled>, { "forbid": [<string>] }]
+"forbid-component-props": [<enabled>, { "forbid": [<string>|<object>], "defaultAllowOnComponents": <bool>, "defaultAllowOnDOM": <bool>}]
````
the two "default" booleans need to be on the object, not on the outer config object. not sure how to denote this tho.

ljharb replied:
unless you're suggesting that they'd be top-level settings as well? if so, i do think i like that.

EdwardDrapkin replied:
yeah, there's an outer config option for specifying the default behavior and an option for overloading the default, but i wasn't quite sure how to make that clear in the documentation other than in the examples...

EdwardDrapkin commented:
Okay, the football sucked, so I went ahead and got to work on this.

I did some extra work that we hadn't discussed, but I was adding examples to the documentation and surfaced some cases I wanted to address.

I added `defaultAllowOnComponents` and `defaultAllowOnDOM` as top level config options to do exactly what they say.  I also allowed a white `allow` alongside the blacklist `forbid` you asked for, because I imagined a case where someone was okay with font icon i tags having classNames but everything else going through something like css modules.

I updated the documentation and I think it's okay, but please do read and update it if you feel it's lacking.  Or tell me and I gladly will!

I also added more test cases so Istanbul reports 100% coverage for me.

Let me know if there's any issues this time.

Thanks so much for your time helping me get this working (and for the rule in the first place)!

ljharb commented:
I think this is looking much better, and I like the additions you've made! I think we'll want to add a very large quantity of test cases, that cover as many interesting combinations of rules as possible.

Commit:
lint errors, my bad

Commit:
Even more test cases

EdwardDrapkin commented:
I added some more test cases, up to 21 now.  I'm having trouble coming up with cases that aren't already tested.

ljharb commented on lib/rules/forbid-component-props.js:
```diff
@@ -22,48 +31,117 @@ module.exports = {
+      {
+        type: 'object',
+        properties: {
+          defaultAllowOnDOM: {
```
i think the schema is missing `defaultAllowOnComponents`?

EdwardDrapkin replied:
oops is all i got

ljharb commented on lib/rules/forbid-component-props.js:
```diff
@@ -22,48 +31,117 @@ module.exports = {
-      additionalProperties: true
-    }]
+        },
+        additionalProperties: true
```
i think we can maybe set this to `false`, so nobody accidentally passes an invalid config option?

EdwardDrapkin replied:
okay, I can change it... I just didn't wanna mess with what you already had

ljharb replied:
yeah that's a good point, it might constitute a breaking change. probably best to leave it `true` for now.

ljharb commented on lib/rules/forbid-component-props.js:
```diff
@@ -22,48 +31,117 @@ module.exports = {
+          defaultAllowOnDOM: {
+            type: 'boolean'
+          },
+          forbid: {
```
the schema also seems to be missing `allow`, which i think should probably share a schema with `forbid`? (ie, reuse the array schema in both)

EdwardDrapkin replied:
okay i was a little excited

ljharb commented on lib/rules/forbid-component-props.js:
```diff
@@ -22,48 +31,117 @@ module.exports = {
 
-      var forbid = configuration.forbid || DEFAULTS;
-      return forbid.indexOf(prop) >= 0;
+      for (var i = 0; i < input.length; i++) {
```
can we use `input.forEach` here, instead of a for loop?

EdwardDrapkin replied:
yep, done

ljharb commented on lib/rules/forbid-component-props.js:
```diff
@@ -22,48 +31,117 @@ module.exports = {
+
+    function getError(prop, tag) {
+      if (rules[prop].forbid.indexOf(tag) > -1) {
+        return 'Prop `' + prop + '` is specifically forbidden on tags named ' + tag;
```
`'… on ' + tag + ' components'`?

EdwardDrapkin replied:
It swaps between "$tag Components" and "$tag DOM nodes" now.

ljharb commented on tests/lib/rules/forbid-component-props.js:
```diff
@@ -143,11 +196,181 @@ ruleTester.run('forbid-component-props', rule, {
+    ]
+  }, {
+    code: [
+      'var First = React.createClass({',
```
for each of the tests you're using for `React.createClass`, let's please also add a test case for an SFC, and a class-based component.

EdwardDrapkin replied:
I only added one total test case for each. I figure if one test case works for all three, all test cases should, because we're only getting at the JSX tags inside the (render) function body, right?  I can add more if you think they're necessary.

ljharb replied:
Ah, that's another thing - we should check for `React.createElement` calls too (ie, non-JSX) :-)

EdwardDrapkin replied:
If you want me to, that's cool and I'm down to learn how, but I don't know... so can you point me in the right direction if there's another plugin I might be able to read?  If not, I'll just RTFM I guess.

And do you have any problem if I do it in a separate pull request?  It seems outside the scope of this one and I might not have time to get to it until this upcoming weekend.

EdwardDrapkin replied:
Hmm, you may have opened a can of worms, at first glance there are several other plugins that do things with props that rely on JSXExpression too... no-duplicate-props, no-unknown-property, sort-props, maybe more....

What is the best way to go forward if this is an issue that affects multiple rules?  Open a new issue and then read each rule one by one and make sure they're not affected?

ljharb replied:
There are a number of rules that address both - however, I'm fine if this rule covers jsx right now, but is expanded to non-JSX in a future PR.

Commit:
fix missing schema elements

Commit:
switch to foreach

Commit:
butter error messages

Commit:
tests for SFC and es6 classes

ljharb commented on lib/rules/forbid-component-props.js:
```diff
@@ -38,6 +45,9 @@ module.exports = {
+          defaultAllowOnComponents: {
+            type: 'boolean'
+          },
           forbid: {
```
is there not a top-level `allow` as well?

EdwardDrapkin replied:
Wouldn't that mean the entire plugin would have to switch into whitelist mode?  Would that be useful to anyone?

ljharb replied:
Hm, yeah i guess if i wanted to override a `forbid` i'd now just modify the object inside the array. This is fine as-is.

Commit:
code readability

Commit:
revert potentially breaking change

EdwardDrapkin commented:
I think I addressed everything from your notes - and I'll make a separate pull request to add `React.createElement` support  - but is there anything else that needs to be done now?

Commit:
Merge remote-tracking branch 'upstream/master'

burabure commented:
this might clash with https://github.com/yannickcr/eslint-plugin-react/pull/936

ljharb commented:
@burabure i think perhaps we might want to follow this PR's lead for "forbidPatterns", and let the "dom or component" options be pattern-specific instead of rule-specific.

burabure commented:
@EdwardDrapkin @ljharb maybe I should take this PR as an starting point for #936 does that sound like a good idea?

ljharb commented:
@EdwardDrapkin would you be alright with @burabure building off your commits to achieve both this PR and #936?

burabure commented:
@EdwardDrapkin ping

burabure commented:
@ljharb I don't think @EdwardDrapkin is going to reply. Should I go ahead?

ljharb commented:
Sure, go for it.

EdwardDrapkin commented:
I'm sorry, I've been really sick lately fighting off MRSA (yikes!) and I won't have time to get to this.  You should definitely go ahead.

ljharb commented:
Thanks for confirming - @burabure, do your thing :-)

burabure commented:
@EdwardDrapkin sorry to hear that, has a friend that had MRSA and I know it really sucks. Hope you feel better soon!

burabure commented:
I don't have time to see this through, if anyone is up to it please do =)
